openedx_lti_tool_plugin/tests/test_views.py

```diff
@@ -143,18 +262,134 @@ def test_with_lti_disabled(self):
+class TestLtiToolLaunchView(LtiViewMixin, TestCase):
     """Test LTI 1.3 platform tool launch view."""
 
     def setUp(self):
         """Test fixtures setup."""
         super().setUp()
         self.url = reverse('lti1p3-launch', args=[COURSE_ID])
         self.view_class = LtiToolLaunchView
-        self.user = MagicMock(username='x', email='x@example.com', is_authenticated=True)
 
-    @patch.object(LtiToolLaunchView, '_authenticate_and_login', return_value=True)
-    @patch.object(CourseKey, 'from_string', return_value=None)
+    @patch.object(LtiProfile.objects, 'get_or_create_from_claims')
+    @patch('openedx_lti_tool_plugin.views.authenticate')
+    @patch('openedx_lti_tool_plugin.views.login')
+    @patch('openedx_lti_tool_plugin.views.mark_user_change_as_expected')
+    def test_authenticate_and_login_with_user(
+        self,
+        mark_user_change_as_expected_mock: MagicMock,
+        login_mock: MagicMock,
+        authenticate_mock: MagicMock,
+        get_or_create_from_claims_mock: MagicMock,
+    ):
+        """Test authenticate_and_login method with user.
+
+        Args:
+            mark_user_change_as_expected_mock: Mocked mark_user_change_as_expected function.
+            login_mock: Mocked login function.
+            authenticate_mock: Mocked authenticate function.
+            get_or_create_from_claims_mock: Mocked LtiProfile get_or_create_from_claims method.
+        """
+        authenticate_mock.return_value = self.user
+
+        self.assertEqual(self.view_class().authenticate_and_login(None, **BASE_LAUNCH_DATA), self.user)
+        get_or_create_from_claims_mock.assert_called_once_with(**BASE_LAUNCH_DATA)
+        authenticate_mock.assert_called_once_with(None, **BASE_LAUNCH_DATA)
+        login_mock.assert_called_once_with(None, self.user)
+        mark_user_change_as_expected_mock.assert_called_once_with(self.user.id)
+
+    @patch.object(LtiProfile.objects, 'get_or_create_from_claims')
+    @patch('openedx_lti_tool_plugin.views.authenticate', return_value=None)
+    def test_authenticate_and_login_without_user(
+        self,
+        authenticate_mock: MagicMock,
+        get_or_create_from_claims_mock: MagicMock,
+    ):
+        """Test authenticate_and_login method without user.
+
+        Args:
+            mark_user_change_as_expected_mock: Mocked mark_user_change_as_expected function.
+            login_mock: Mocked login function.
+            authenticate_mock: Mocked authenticate function.
+            get_or_create_from_claims_mock: Mocked LtiProfile get_or_create_from_claims method.
+        """
+        self.assertEqual(self.view_class().authenticate_and_login(None, **BASE_LAUNCH_DATA), None)
+        get_or_create_from_claims_mock.assert_called_once_with(**BASE_LAUNCH_DATA)
+        authenticate_mock.assert_called_once_with(None, **BASE_LAUNCH_DATA)
+
+    @patch('openedx_lti_tool_plugin.views.course_enrollment')
+    def test_enroll_with_enrollment(self, course_enrollment_mock: MagicMock):
```

Comment:
Check that course_enrollment().enroll has not been called. @kuipumu Could you be explicit in what the course_enrollment mock return? in this case True.

Reply:
@Squirrel18 Do you mean course_enrollment_mock().get_enrollment return?, it should return the enrollment object or None, not a boolean